spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/UpdatableStreamService.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.dataflow.server.service;
+
+import java.util.Collection;
+
+import org.springframework.cloud.dataflow.rest.UpdateStreamRequest;
+import org.springframework.cloud.skipper.domain.Deployer;
+import org.springframework.cloud.skipper.domain.Release;
+
+/**
+ * Extends the core {@link StreamService} operations with support for update, rollback and deployment history
+ *
+ * @author Mark Pollack
+ * @author Ilayaperumal Gopinathan
+ * @author Christian Tzolov
+ */
+public interface UpdatableStreamService extends StreamService {
+
+	/**
+	 * Update the stream using the UpdateStreamRequest.
+	 *
+	 * @param streamName the name of the stream to update
+	 * @param updateStreamRequest the UpdateStreamRequest to use during the update
+	 */
+	void updateStream(String streamName, UpdateStreamRequest updateStreamRequest);
+
+	/**
+	 * Rollback the stream to the previous or a specific version of the stream.
+	 *
+	 * @param streamName the name of the stream to rollback
+	 * @param releaseVersion the version to rollback to (if not specified, rollback to the previous deleted/deployed
+	 * release version of the stream.
+	 */
+	void rollbackStream(String streamName, int releaseVersion);
+
+	/**
+	 * Return a manifest info of a release version. For packages with dependencies, the
+	 * manifest includes the contents of those dependencies.
+	 *
+	 * @param releaseName the release name
+	 * @param releaseVersion the release version
+	 * @return the manifest info of a release
+	 */
+	String manifest(String releaseName, int releaseVersion);
```

Comment:
While `updateStream` and `rollbackStream` methods sound generic enough, the other methods can possibly moved into another interface that extends `UpdatableStreamService`. This way, if we have another mode that can just do `update` and `rollback` those implementations could use `UpdatableStreamService` directly.

Reply:
This is a good point and i completely agree. 
As it stands at the moment though to do this we need also to change the updateStream method signature to replace the skipper specific UpdateRequest with something generic. 
Perhaps we can postpone this fine tuning once we add a new mode to integrate?

Reply:
It isn't clear if the other methods such as `manifest` and `platformList` would apply to anything else other than when delegating to a skipper backend.  Perhaps a better name for this interface is `SkipperStreamService` and the current implementation - `public class SkipperStreamService extends AbstractStreamService implements UpdatableStreamService` changes to

`public class DefaultSkipperStreamService extends AbstractStreamService implements SkipperStreamService`

Reply:
@markpollack , sure this will make the naming more consistent. Will rename the interfaces and classes and will update the PR.   
Btw what do you think about this package structure ?
- org.springframework.cloud.dataflow.server.service.*
- org.springframework.cloud.dataflow.server.service.impl.*
- org.springframework.cloud.dataflow.server.stream.*

Why the `StreamService` is in the `...server.service` while the `StreamDeployer` is inside the `...server.stream` packages?

Reply:
The structure isn't bad from the DSM/structure101 perspective, not really and tangles, the `server.stream` package sits at the bottom of the pile.  The name of the package would be better off as `server.deployer` instead of `server.stream` but we should handle this as part of a general tech-debt story and take a look at it inside structure101 to get a complete picture.  I'll create an issue if we don't already have one on this topic.

Reply:
Created https://github.com/spring-cloud/spring-cloud-dataflow/issues/2054